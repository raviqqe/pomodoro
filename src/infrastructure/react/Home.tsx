import React from "react";
import { PulseLoader } from "react-spinners";
import styled from "styled-components";
import { SignOut } from "./SignOut";

const Container = styled.div`
  display: flex;
  flex-direction: column;
  height: 100vh;
  width: 85ex;
  max-width: 100%;
  margin: auto;

  > :first-child {
    flex: 1;
  }
`;

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`;

const SignOutContainer = styled.div`
  position: fixed;
  top: 0.5em;
  right: 0.5em;
`;

export interface IProps {
  signOut: () => Promise<void>;
}

export const Home = ({ signOut }: IProps) => {
  return (
    <Container>
      <LoaderContainer>
        <PulseLoader color="white" />
      </LoaderContainer>
      <SignOutContainer>
        <SignOut signOut={signOut} />
      </SignOutContainer>
    </Container>
  );
};
