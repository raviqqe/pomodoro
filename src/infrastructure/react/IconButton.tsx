import React, { AriaAttributes, ReactNode } from "react";
import styled from "styled-components";

const Button = styled.div`
  text-shadow: 0rem 0.2rem 0.2rem rgba(0, 0, 0, 0.2);
  color: grey;
  cursor: pointer;
  font-size: 1.5em;
  display: flex;
`;

interface IProps extends AriaAttributes {
  children: ReactNode;
  onClick: () => void;
}

export const IconButton = ({ children, onClick, ...restProps }: IProps) => (
  <Button onClick={onClick} {...restProps}>
    {children}
  </Button>
);
