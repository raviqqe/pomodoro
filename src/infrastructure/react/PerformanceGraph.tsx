import { last } from "lodash";
import { DateTime } from "luxon";
import {
  CartesianGrid,
  ResponsiveContainer,
  XAxis,
  YAxis,
  Label,
  Bar,
  BarChart,
} from "recharts";
import styled from "styled-components";
import { IPerformanceGraph } from "../../application/performance-graph";
import { DateSerializer } from "../../domain/date-serializer";
import { white } from "./style/colors";

const Container = styled.div`
  width: 80vw;
  height: 80vh;
`;

const Message = styled.div`
  color: ${white};
  font-size: 1.5em;
`;

export interface IProps {
  performanceGraph: IPerformanceGraph;
}

export const PerformanceGraph = ({
  performanceGraph: { data },
}: IProps): JSX.Element =>
  data.length === 0 ? (
    <Message>No performance graph to show yet!</Message>
  ) : (
    <Container>
      <ResponsiveContainer>
        <BarChart data={data}>
          <CartesianGrid fill="white" stroke="grey" strokeDasharray="3 3" />
          <XAxis
            dataKey="date"
            stroke="grey"
            tickFormatter={(date: string): string => {
              const days: number = DateTime.fromJSDate(
                DateSerializer.deserialize(
                  (last(data) as { date: string }).date
                )
              ).diff(
                DateTime.fromJSDate(DateSerializer.deserialize(date)),
                "days"
              ).days;

              return days === 0 ? "Today" : `${days} days ago`;
            }}
            tickMargin={10}
          />
          <YAxis allowDecimals={false} stroke="grey" tickMargin={5}>
            <Label
              angle={-90}
              position="insideLeft"
              style={{ fill: "grey" }}
              value="Pomodoros"
            />
          </YAxis>
          <Bar dataKey="pomodoros" fill="salmon" />
        </BarChart>
      </ResponsiveContainer>
    </Container>
  );
