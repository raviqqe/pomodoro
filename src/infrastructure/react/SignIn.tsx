import React from "react";
import { TextButton } from "./TextButton";

interface IProps {
  signIn: () => void;
}

export const SignIn = ({ signIn }: IProps) => (
  <TextButton onClick={signIn}>Sign in</TextButton>
);
