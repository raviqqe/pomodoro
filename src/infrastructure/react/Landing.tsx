import { GoMarkGithub } from "react-icons/go";
import React from "react";
import styled from "styled-components";
import { SignIn } from "./SignIn";

const Container = styled.div`
  height: 100vh;
  width: 100vw;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  > * {
    margin: 2rem;
  }
`;

const Title = styled.div`
  color: steelblue;
  font-family: "Chelsea Market", sans-serif;
  font-size: 4em;
  font-weight: bold;
  letter-spacing: -0.35ex;
  text-shadow: 0.4em 0.4em 0.4em rgba(0, 0, 0, 0.1);
`;

const GitHubLink = styled.a`
  font-size: 2.5rem;
  margin: 1.5rem;
  color: #222;
  display: block;
  position: fixed;
  bottom: 0em;
  right: 0em;
  line-height: 0ex;
`;

export interface IProps {
  repositoryURL: string;
  signIn: () => void;
}

export const Landing = ({ repositoryURL, signIn }: IProps) => (
  <Container>
    <Title>P🍅m🍅d🍅r🍅</Title>
    <SignIn signIn={signIn} />
    <GitHubLink href={repositoryURL} target="_blank">
      <GoMarkGithub />
    </GitHubLink>
  </Container>
);
