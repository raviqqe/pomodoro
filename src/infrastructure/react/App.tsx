import { Omit } from "lodash";
import { PulseLoader } from "react-spinners";
import { useAsync } from "react-use";
import React, { useState } from "react";
import styled from "styled-components";
import { IInitialState } from "../../application/application-initializer";
import { IProps as ILandingProps, Landing } from "./Landing";
import { Home, IProps as IHomeProps } from "./Home";

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  width: 100vw;
`;

interface IProps extends Omit<IHomeProps, "signOut">, ILandingProps {
  initialize: () => Promise<IInitialState>;
  signIn: () => Promise<boolean>;
  signOut: () => Promise<boolean>;
}

export const App = ({
  initialize,
  repositoryURL,
  signIn,
  signOut,
  ...props
}: IProps) => {
  const [signedIn, setSignedIn] = useState<boolean | null>(null);

  useAsync(async () => {
    const { signedIn } = await initialize();
    setSignedIn(signedIn);
  }, []);

  if (signedIn === null) {
    return (
      <LoaderContainer>
        <PulseLoader color="white" />
      </LoaderContainer>
    );
  } else if (!signedIn) {
    return (
      <Landing
        repositoryURL={repositoryURL}
        signIn={async () => setSignedIn(await signIn())}
      />
    );
  }

  return <Home {...props} signOut={async () => setSignedIn(await signOut())} />;
};
