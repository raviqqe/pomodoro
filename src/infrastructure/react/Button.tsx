import { FC, HTMLAttributes, HTMLProps, ReactNode } from "react";
import styled from "styled-components";
import { boxShadow } from "./style";

interface IProps
  extends HTMLProps<HTMLButtonElement>,
    HTMLAttributes<HTMLButtonElement> {
  secondary?: boolean;
  children?: ReactNode;
}

export const Button: FC<IProps> = styled.button<{ secondary?: boolean }>`
  ${boxShadow};
  background: ${({ secondary }) => (secondary ? "darkgrey" : "salmon")}
  display: flex;
  justify-content: center;
  align-items: center;
  font-family: inherit;
  color: white;
  border: none;
  cursor: pointer;
  flex-shrink: 0;
` as FC<any>;
