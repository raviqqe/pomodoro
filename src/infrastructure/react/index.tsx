import { render } from "react-dom";
import React from "react";
import { ApplicationInitializer } from "../../application/application-initializer";
import { SignInManager } from "../../application/sign-in-manager";
import { SignOutManager } from "../../application/sign-out-manager";
import { PerformanceGraphViewer } from "../../application/performance-graph-viewer";
import { PomodoroTimerStopper } from "../../application/pomodoro-timer-stopper";
import { PomodoroTimerStarter } from "../../application/pomodoro-timer-starter";
import { PomodoroTimerState } from "../../application/pomodoro-timer-state";
import { IRenderer, IPomodoroTimer } from "../renderer";
import { IPerformanceGraph } from "../../application/performance-graph";
import { App, IProps as IAppProps } from "./App";
import { GlobalStyle } from "./style";

interface IPresenter {
  setRenderer(renderer: IRenderer): void;
}

interface IProps
  extends Pick<IAppProps, "performanceGraph" | "signedIn" | "timer"> {}

export class ReactRenderer implements IRenderer {
  private props: IProps = {
    performanceGraph: { data: [] },
    signedIn: null,
    timer: { seconds: 0, state: PomodoroTimerState.Pomodoro, stopped: true },
  };

  constructor(
    private readonly element: HTMLElement,
    presenters: IPresenter[],
    private readonly applicationInitializer: ApplicationInitializer,
    private readonly performanceGraphViewer: PerformanceGraphViewer,
    private readonly pomodoroTimerStarter: PomodoroTimerStarter,
    private readonly pomodoroTimerStopper: PomodoroTimerStopper,
    private readonly signInManager: SignInManager,
    private readonly signOutManager: SignOutManager,
    private readonly repositoryURL: string
  ) {
    for (const presenter of presenters) {
      presenter.setRenderer(this);
    }
  }

  public render(): void {
    this.renderProps({});
  }

  public renderPerformanceGraph(performanceGraph: IPerformanceGraph): void {
    this.renderProps({ performanceGraph });
  }

  public renderSignedIn(signedIn: boolean): void {
    this.renderProps({ signedIn });
  }

  public renderTimer(timer: IPomodoroTimer): void {
    this.renderProps({ timer });
  }

  private renderProps(props: Partial<IProps>): void {
    this.props = { ...this.props, ...props };

    render(
      <>
        <App
          {...this.props}
          initialize={() => this.applicationInitializer.initialize()}
          repositoryURL={this.repositoryURL}
          signIn={() => this.signInManager.signIn()}
          signOut={() => this.signOutManager.signOut()}
          startTimer={() => this.pomodoroTimerStarter.start()}
          stopTimer={() => this.pomodoroTimerStopper.stop()}
          viewGraph={() => this.performanceGraphViewer.viewGraph()}
        />
        <GlobalStyle />
      </>,
      this.element
    );
  }
}
