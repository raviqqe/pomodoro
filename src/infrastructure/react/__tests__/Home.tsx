import { render, fireEvent, waitFor } from "@testing-library/react";
import { create } from "react-test-renderer";
import { PomodoroTimerState } from "../../../application/pomodoro-timer-state";
import { Home, IProps } from "../Home";

const props: IProps = {
  performanceGraph: { data: [] },
  signOut: async () => {},
  startTimer: async () => {},
  stopTimer: async () => {},
  timer: { seconds: 42, state: PomodoroTimerState.Pomodoro, stopped: false },
  viewGraph: async () => {},
};

it("renders", () => {
  expect(create(<Home {...props} />).toJSON()).toMatchSnapshot();
});

it("views a performance graph", async () => {
  const { container } = render(<Home {...props} />);

  fireEvent.click(
    container.querySelector(`[aria-label="View Graph"]`) as Element
  );
  await waitFor(() => container.querySelector(`[aria-label="View Timer"]`));

  expect(container).toMatchSnapshot();
});

it("goes back to a timer view", async () => {
  const { container } = render(<Home {...props} />);

  fireEvent.click(
    container.querySelector(`[aria-label="View Graph"]`) as Element
  );
  await waitFor(() => container.querySelector(`[aria-label="View Timer"]`));

  fireEvent.click(
    container.querySelector(`[aria-label="View Timer"]`) as Element
  );

  expect(container).toMatchSnapshot();
});
