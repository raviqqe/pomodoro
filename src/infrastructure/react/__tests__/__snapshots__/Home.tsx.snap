// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`goes back to a timer view 1`] = `
<div>
  <div
    class="sc-fzXfLP iXndoa"
  >
    <div
      class="sc-AykKF kQKBf"
    >
      <div
        class="sc-AykKH cIXvBs"
      >
        <span
          class="sc-AykKI fSiutg"
        >
          0
        </span>
        <span
          class="sc-AykKJ hguCqt"
        >
          42
        </span>
      </div>
      <button
        class="sc-AykKC sc-AykKE cJtYbK"
      >
        Stop
      </button>
    </div>
    <div
      class="sc-fzXfLQ jJLvIq"
    >
      <button
        class="sc-AykKC sc-AykKD fyAhDl"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </button>
      <button
        aria-label="View Graph"
        class="sc-AykKC sc-AykKD fyAhDl"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M176 64h64v384h-64zM80 336h64v112H80zM272 272h64v176h-64zM368 176h64v272h-64z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`renders 1`] = `
<div
  className="sc-fzXfLP iXndoa"
>
  <div
    className="sc-AykKF kQKBf"
  >
    <div
      className="sc-AykKH cIXvBs"
    >
      <span
        className="sc-AykKI fSiutg"
      >
        0
      </span>
      <span
        className="sc-AykKJ hguCqt"
      >
        42
      </span>
    </div>
    <button
      className="sc-AykKC sc-AykKE cJtYbK"
      onClick={[Function]}
    >
      Stop
    </button>
  </div>
  <div
    className="sc-fzXfLQ jJLvIq"
  >
    <button
      className="sc-AykKC sc-AykKD fyAhDl"
      onClick={[Function]}
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
    </button>
    <button
      aria-label="View Graph"
      className="sc-AykKC sc-AykKD fyAhDl"
      onClick={[Function]}
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M176 64h64v384h-64zM80 336h64v112H80zM272 272h64v176h-64zM368 176h64v272h-64z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`views a performance graph 1`] = `
<div>
  <div
    class="sc-fzXfLP iXndoa"
  >
    <div
      class="sc-fzXfLO kdSBrU"
    >
      No performance graph to show yet!
    </div>
    <div
      class="sc-fzXfLQ jJLvIq"
    >
      <button
        class="sc-AykKC sc-AykKD fyAhDl"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </button>
      <button
        aria-label="View Timer"
        class="sc-AykKC sc-AykKD fyAhDl"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15 1H9v2h6V1zm-4 13h2V8h-2v6zm8.03-6.61l1.42-1.42c-.43-.51-.9-.99-1.41-1.41l-1.42 1.42C16.07 4.74 14.12 4 12 4c-4.97 0-9 4.03-9 9s4.02 9 9 9 9-4.03 9-9c0-2.12-.74-4.07-1.97-5.61zM12 20c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
